crates/bevy_core_pipeline/src/bloom/bloom.wgsl

```diff
@@ -1,138 +1,137 @@
-    let d = textureSample(original, original_sampler, uv + vec2<f32>(-0.5, -0.5) * scale);
-    let e = textureSample(original, original_sampler, uv + vec2<f32>(0.5, -0.5) * scale);
-    let f = textureSample(original, original_sampler, uv + vec2<f32>(-1.0, 0.0) * scale);
-    let g = textureSample(original, original_sampler, uv + vec2<f32>(0.0, 0.0) * scale);
-    let h = textureSample(original, original_sampler, uv + vec2<f32>(1.0, 0.0) * scale);
-    let i = textureSample(original, original_sampler, uv + vec2<f32>(-0.5, 0.5) * scale);
-    let j = textureSample(original, original_sampler, uv + vec2<f32>(0.5, 0.5) * scale);
-    let k = textureSample(original, original_sampler, uv + vec2<f32>(-1.0, 1.0) * scale);
-    let l = textureSample(original, original_sampler, uv + vec2<f32>(0.0, 1.0) * scale);
-    let m = textureSample(original, original_sampler, uv + vec2<f32>(1.0, 1.0) * scale);
-
-    let div = (1.0 / 4.0) * vec2<f32>(0.5, 0.125);
-
-    var o: vec4<f32> = (d + e + i + j) * div.x;
-    o = o + (a + b + g + f) * div.y;
-    o = o + (b + c + h + g) * div.y;
-    o = o + (f + g + l + k) * div.y;
-    o = o + (g + h + m + l) * div.y;
-
-    return o;
+// "Next Generation Post Processing in Call of Duty: Advanced Warfare" slide 100
+fn sample_input_13_tap(uv: vec2<f32>) -> vec3<f32> {
+    let texel_size = 1.0 / vec2<f32>(textureDimensions(input_texture));
+    let x = texel_size.x;
+    let y = texel_size.y;
+
+    let a = textureSample(input_texture, s, vec2<f32>(uv.x - 2.0 * x, uv.y + 2.0 * y)).rgb;
+    let b = textureSample(input_texture, s, vec2<f32>(uv.x, uv.y + 2.0 * y)).rgb;
+    let c = textureSample(input_texture, s, vec2<f32>(uv.x + 2.0 * x, uv.y + 2.0 * y)).rgb;
+
+    let d = textureSample(input_texture, s, vec2<f32>(uv.x - 2.0 * x, uv.y)).rgb;
+    let e = textureSample(input_texture, s, vec2<f32>(uv.x, uv.y)).rgb;
+    let f = textureSample(input_texture, s, vec2<f32>(uv.x + 2.0 * x, uv.y)).rgb;
+
+    let g = textureSample(input_texture, s, vec2<f32>(uv.x - 2.0 * x, uv.y - 2.0 * y)).rgb;
+    let h = textureSample(input_texture, s, vec2<f32>(uv.x, uv.y - 2.0 * y)).rgb;
+    let i = textureSample(input_texture, s, vec2<f32>(uv.x + 2.0 * x, uv.y - 2.0 * y)).rgb;
+
+    let j = textureSample(input_texture, s, vec2<f32>(uv.x - x, uv.y + y)).rgb;
+    let k = textureSample(input_texture, s, vec2<f32>(uv.x + x, uv.y + y)).rgb;
+    let l = textureSample(input_texture, s, vec2<f32>(uv.x - x, uv.y - y)).rgb;
+    let m = textureSample(input_texture, s, vec2<f32>(uv.x + x, uv.y - y)).rgb;
+
+#ifdef FIRST_DOWNSAMPLE
+    // "Next Generation Post Processing in Call of Duty: Advanced Warfare" slide 112
+    var group0 = (a + b + d + e) * (0.125f / 4.0f);
+    var group1 = (b + c + e + f) * (0.125f / 4.0f);
+    var group2 = (d + e + g + h) * (0.125f / 4.0f);
+    var group3 = (e + f + h + i) * (0.125f / 4.0f);
+    var group4 = (j + k + l + m) * (0.5f / 4.0f);
+    group0 *= karis_average(group0);
```

Comment:
The article says:

> This operation, when used to reduce fireflies, can also be thought of as a weighting function for each sample:
> `weight = 1 / (1 + luma)`
> 
> For a weighted average, sum all samples and divide by the summed weights. The result will be the same as if the samples were tone mapped using T with range of 1, averaged, then inverse tone mapped using Tinverse.

I am not convinced the implementation in this PR is a valid simplification of that. For one, I don't think calculating the weight of a sum of samples will be the same as summing the weights of samples, when the formula for the weight is `1 / (1 + luma)`. `(1 / a) + (1 / b) = (b / (a * b)) + (a / (a * b)) = (a + b) / (a * b)`.

I would say that I think to be correct, the firefly reduction would need to be applied to each sample individually. But then I realised that each sample is a bilinear interpolation of samples around it, so it is already in some sense a weighted-average value.

So, what is the goal here? To reduce fireflies. The PR considers each group, which are groups of four samples in various configurations at least 2 pixels apart, and then tries to eliminate fireflies in the average of those four. It seems incorrect regardless.

I'm not yet sure what a correct and/or acceptable solution is.

Reply:
It seems I missed part of the slides that discuss this approach. It was intentional to apply the firefly reduction to the averaged groups of 4 samples, so this is OK.

We should add a reference for it though. Had it been there I'd have looked at it. :)

Reply:
The firefly reduction method used in this PR is a well tested one that is extensively described and, as far as I see, correctly implemented one. There is another [article](https://learnopengl.com/Guest-Articles/2022/Phys.-Based-Bloom) (Bonus material 2: Karis average) that shows code very similar to what we have

Reply:
Thanks for that reference. We should probably add a link to it too.

Reply:
I see now there is a link at the top of the file.

Reply:
Clarity: I'm happy to consider this resolved.